ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/PrimitiveIm2Col.kt

```diff
@@ -0,0 +1,140 @@
+@file:GeneratePrimitives(DataType.FLOAT, DataType.DOUBLE)
+
+package io.kinference.ndarray.extensions
+
+import io.kinference.ndarray.arrays.PrimitiveNDArray
+import io.kinference.ndarray.arrays.pointers.PrimitivePointer
+import io.kinference.ndarray.arrays.tiled.PrimitiveTiledArray
+import io.kinference.primitives.annotations.GeneratePrimitives
+import io.kinference.primitives.annotations.SpecifyPrimitives
+import io.kinference.primitives.types.*
+
+fun primitiveIm2Col(
+    im: PrimitivePointer,
+    imageShape: IntArray,
+    outputShape: IntArray,
+    channelsCol: Int,
+    kernelShape: IntArray,
+    stride: IntArray,
+    dilation: IntArray,
+    pad: IntArray,
+    rank: Int,
+    col: PrimitiveArray,
+    padValue: PrimitiveType = 0.toPrimitive()
+) {
+    val kernelSize = kernelShape.shapeSize()
+    val dOffset = IntArray(rank)
+    val dIter = IntArray(rank)
+
+    for (cCol in 0 until channelsCol) {
+        var offset = cCol
+        for (dI in rank - 1 downTo 0) {
+            if (dI < rank - 1) {
+                offset /= kernelShape[dI + 1]
+            }
+            dOffset[dI] = offset % kernelShape[dI]
+        }
+
+        var hasNextOutput: Boolean
+        do {
+            var indexCol = cCol
+            var indexIm = cCol / kernelSize
+            var isPadding = false
+
+            for (dI in 0 until rank) {
+                val d = dIter[dI]
+                val dIm = d * stride[dI] - pad[dI] + dOffset[dI] * dilation[dI]
+                if (dIm < 0 || dIm >= imageShape[dI])
+                    isPadding = true
+
+                indexCol *= outputShape[dI]
+                indexCol += d
+
+                indexIm *= imageShape[dI]
+                indexIm += dIm
+            }
+
+            if (isPadding)
+                col[indexCol] = padValue
+            else
+                col[indexCol] = im.array[indexIm + im.linearIndex]
+
+            hasNextOutput = false
+            for (j in rank - 1 downTo 0) {
+                val dMax = outputShape[j] - 1
+                if (dIter[j] >= dMax) {
+                    dIter[j] = 0
+                } else {
+                    ++dIter[j]
+                    hasNextOutput = true
+                    break
+                }
+            }
+        } while (hasNextOutput)
+    }
+}
+
+@SpecifyPrimitives(include = [])
+fun isInPadding(actual: Int, bound: Int) : Boolean {
+    return actual < 0 || actual >= bound
+}
+
+fun primitiveIm2Col(
```

Comment:
Maybe some data classes would work for this? Smth like PadInfo and Height and Width info for example